templates/vue/src/components/TapestryMain/RootNodeButton/index.vue

```diff
@@ -88,88 +93,90 @@ export default {
       evt.preventDefault()
       evt.stopPropagation()
       const file = evt.dataTransfer.files[0]
-      if (!file.name.endsWith("json")) {
-        this.error = {
-          message: "Please upload a JSON file.",
-        }
-      } else {
-        this.importTapestry(file)
-      }
+      this.importTapestryFromFile(file)
     },
     handleFileChange() {
       const file = this.$refs.fileInput.files[0]
-      this.importTapestry(file)
+      this.importTapestryFromFile(file)
+    },
+    importTapestryFromFile(file) {
+      if (!file) return
+
+      switch (this.getFileExtension(file.name)) {
+        case "json":
+          this.error = null
+          this.importTapestry(file)
+          break
+        case "zip":
+          this.error = null
+          this.importTapestryFromZip(file)
+          break
+        default:
+          this.error = {
+            message: "Please upload a JSON file or a ZIP file.",
+          }
+      }
     },
     importTapestry(file) {
       const reader = new FileReader()
       reader.onload = async e => {
         this.error = ""
 
         this.isImporting = true
-        let upload
-        try {
-          upload = JSON.parse(e.target.result)
-          this.validateTapestryJSON(upload)
-        } catch (err) {
-          this.error = err
-          this.isImporting = false
-          return
-        }
-        if (!(upload["site-url"] == wpData.wpUrl)) {
-          await this.prepareImport(upload)
-        }
+        const upload = e.target.result
         client
           .importTapestry(upload)
-          .then(() => {
-            this.isImporting = false
+          .then(response => {
+            this.changes.permissions = new Set(response.changes.permissions)
```

Comment:
Yes, the backend can return duplicate permission changes. This code is run for every permissions object we filter, so it may add the same role twice if that role was stripped from multiple permissions objects.

https://github.com/tapestry-tool/tapestry-wp/blob/c54bcf759b8f4c727c7612fb1077f3924800271e/utilities/class.tapestry-import-export.php#L87-L92

I think this property was originally a Set for the same reason. However, if it would be better, I can change it to remove duplicates in the backend and use an array here.

Reply:
I don't think there's a need to do it in the backend, so let's just keep it as-is for now.